File: android/src/main/kotlin/com/flueraengine/fluera_engine/AudioPlayerPlugin.kt
package com.flueraengine.fluera_engine

import android.content.Context
import android.media.MediaPlayer
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.File

/**
 * 🎵 AudioPlayerPlugin — Native audio playback for Fluera Engine.
 *
 * Handles playback of recorded audio files using Android MediaPlayer.
 *
 * Platform Channel: flueraengine.audio/player
 * Event Channel: flueraengine.audio/player_events
 */
class AudioPlayerPlugin : FlutterPlugin, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var context: Context? = null

    private var mediaPlayer: MediaPlayer? = null
    private var eventSink: EventChannel.EventSink? = null
    private var updateHandler: Handler? = null
    private var updateRunnable: Runnable? = null
    private var isPrepared = false

    // MARK: - FlutterPlugin

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, "flueraengine.audio/player")
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, "flueraengine.audio/player_events")
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        releasePlayer()
        context = null
    }

    // MARK: - MethodCallHandler

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> handleInitialize(result)
            "setFilePath" -> {
                val path = call.argument<String>("path")
                if (path != null) handleSetFilePath(path, result)
                else result.error("INVALID_ARGS", "path is required", null)
            }
            "play" -> handlePlay(result)
            "pause" -> handlePause(result)
            "stop" -> handleStop(result)
            "seek" -> {
                val position = call.argument<Int>("position") ?: 0
                handleSeek(position, result)
            }
            "setVolume" -> {
                val volume = call.argument<Double>("volume") ?: 1.0
                handleSetVolume(volume.toFloat(), result)
            }
            "setSpeed" -> {
                val speed = call.argument<Double>("speed") ?: 1.0
                handleSetSpeed(speed.toFloat(), result)
            }
            "getPosition" -> handleGetPosition(result)
            "getDuration" -> handleGetDuration(result)
            "getState" -> handleGetState(result)
            "release" -> {
                releasePlayer()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    // MARK: - EventChannel.StreamHandler

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    // MARK: - Player Methods

    private fun handleInitialize(result: MethodChannel.Result) {
        result.success(null)
    }

    private fun handleSetFilePath(path: String, result: MethodChannel.Result) {
        try {
            releasePlayer()

            val file = File(path)
            if (!file.exists()) {
                result.error("FILE_NOT_FOUND", "File not found: $path", null)
                return
            }

            mediaPlayer = MediaPlayer().apply {
                setDataSource(path)

                setOnCompletionListener {
                    stopPositionUpdates()
                    sendStateEvent("completed")
                }

                setOnErrorListener { _, what, extra ->
                    isPrepared = false
                    sendEvent("error", mapOf("error" to "MediaPlayer error: what=$what extra=$extra"))
                    true
                }

                // Use synchronous prepare for local files — fast and avoids race conditions
                prepare()
            }

            isPrepared = true
            sendStateEvent("ready")
            sendEvent("duration", mapOf("duration" to mediaPlayer!!.duration))
            result.success(null)
        } catch (e: Exception) {
            result.error("SET_FILE_FAILED", e.message, null)
        }
    }

    private fun handlePlay(result: MethodChannel.Result) {
        val player = mediaPlayer
        if (player == null || !isPrepared) {
            result.error("NOT_READY", "Player not ready", null)
            return
        }
        try {
            player.start()
            startPositionUpdates()
            sendStateEvent("playing")
            result.success(null)
        } catch (e: Exception) {
            result.error("PLAY_FAILED", e.message, null)
        }
    }

    private fun handlePause(result: MethodChannel.Result) {
        val player = mediaPlayer
        if (player == null || !isPrepared) {
            result.error("NOT_READY", "Player not ready", null)
            return
        }
        try {
            player.pause()
            stopPositionUpdates()
            sendStateEvent("paused")
            result.success(null)
        } catch (e: Exception) {
            result.error("PAUSE_FAILED", e.message, null)
        }
    }

    private fun handleStop(result: MethodChannel.Result) {
        val player = mediaPlayer
        if (player == null || !isPrepared) {
            result.success(null)
            return
        }
        try {
            player.stop()
            stopPositionUpdates()
            player.prepareAsync()
            sendStateEvent("stopped")
            result.success(null)
        } catch (e: Exception) {
            result.error("STOP_FAILED", e.message, null)
        }
    }

    private fun handleSeek(positionMs: Int, result: MethodChannel.Result) {
        val player = mediaPlayer
        if (player == null || !isPrepared) {
            result.error("NOT_READY", "Player not ready", null)
            return
        }
        try {
            player.seekTo(positionMs)
            result.success(null)
        } catch (e: Exception) {
            result.error("SEEK_FAILED", e.message, null)
        }
    }

    private fun handleSetVolume(volume: Float, result: MethodChannel.Result) {
        try {
            mediaPlayer?.setVolume(volume, volume)
            result.success(null)
        } catch (e: Exception) {
            result.error("VOLUME_FAILED", e.message, null)
        }
    }

    private fun handleSetSpeed(speed: Float, result: MethodChannel.Result) {
        val player = mediaPlayer
        if (player == null || !isPrepared) {
            result.error("NOT_READY", "Player not ready", null)
            return
        }
        try {
            val params = player.playbackParams.setSpeed(speed)
            player.playbackParams = params
            result.success(null)
        } catch (e: Exception) {
            result.error("SPEED_FAILED", e.message, null)
        }
    }

    private fun handleGetPosition(result: MethodChannel.Result) {
        val player = mediaPlayer
        if (player == null || !isPrepared) {
            result.success(0)
            return
        }
        result.success(player.currentPosition)
    }

    private fun handleGetDuration(result: MethodChannel.Result) {
        val player = mediaPlayer
        if (player == null || !isPrepared) {
            result.success(null)
            return
        }
        result.success(player.duration)
    }

    private fun handleGetState(result: MethodChannel.Result) {
        val player = mediaPlayer
        val state = when {
            player == null -> "idle"
            !isPrepared -> "loading"
            player.isPlaying -> "playing"
            else -> "paused"
        }
        result.success(mapOf("state" to state))
    }

    // MARK: - Position Updates

    private fun startPositionUpdates() {
        stopPositionUpdates()
        updateHandler = Handler(Looper.getMainLooper())
        updateRunnable = object : Runnable {
            override fun run() {
                val player = mediaPlayer ?: return
                if (isPrepared && player.isPlaying) {
                    sendEvent("position", mapOf("position" to player.currentPosition))
                    updateHandler?.postDelayed(this, 200)
                }
            }
        }
        updateHandler?.post(updateRunnable!!)
    }

    private fun stopPositionUpdates() {
        updateRunnable?.let { updateHandler?.removeCallbacks(it) }
        updateRunnable = null
        updateHandler = null
    }

    // MARK: - Event Sending

    private fun sendStateEvent(state: String) {
        sendEvent("state", mapOf("state" to state))
    }

    private fun sendEvent(type: String, data: Map<String, Any?>) {
        val event = mutableMapOf<String, Any?>("event" to type)
        event.putAll(data)
        Handler(Looper.getMainLooper()).post {
            eventSink?.success(event)
        }
    }

    // MARK: - Release

    private fun releasePlayer() {
        stopPositionUpdates()
        try {
            mediaPlayer?.release()
        } catch (_: Exception) {}
        mediaPlayer = null
        isPrepared = false
    }
}

File: ios/Classes/AudioPlayerPlugin.swift
import Flutter
import AVFoundation

/// 🎵 AudioPlayerPlugin — Native audio playback for Fluera Engine (iOS).
///
/// Handles playback of recorded audio files using AVAudioPlayer.
///
/// Platform Channel: `flueraengine.audio/player`
/// Event Channel: `flueraengine.audio/player_events`
public class AudioPlayerPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?
    private var audioPlayer: AVAudioPlayer?
    private var positionTimer: Timer?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = AudioPlayerPlugin()
        let channel = FlutterMethodChannel(
            name: "flueraengine.audio/player",
            binaryMessenger: registrar.messenger()
        )
        let eventChannel = FlutterEventChannel(
            name: "flueraengine.audio/player_events",
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: channel)
        eventChannel.setStreamHandler(instance)
        instance.methodChannel = channel
        instance.eventChannel = eventChannel
    }



    // MARK: - FlutterStreamHandler

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }

    // MARK: - MethodCallHandler

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            // Configure audio session for playback
            do {
                try AVAudioSession.sharedInstance().setCategory(.playback, mode: .default)
                try AVAudioSession.sharedInstance().setActive(true)
                result(nil)
            } catch {
                result(FlutterError(code: "INIT_FAILED", message: error.localizedDescription, details: nil))
            }

        case "setFilePath":
            guard let args = call.arguments as? [String: Any],
                  let path = args["path"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "path is required", details: nil))
                return
            }
            handleSetFilePath(path, result: result)

        case "play":
            handlePlay(result: result)

        case "pause":
            handlePause(result: result)

        case "stop":
            handleStop(result: result)

        case "seek":
            guard let args = call.arguments as? [String: Any],
                  let positionMs = args["position"] as? Int else {
                result(FlutterError(code: "INVALID_ARGS", message: "position is required", details: nil))
                return
            }
            handleSeek(positionMs: positionMs, result: result)

        case "setVolume":
            guard let args = call.arguments as? [String: Any],
                  let volume = args["volume"] as? Double else {
                result(FlutterError(code: "INVALID_ARGS", message: "volume is required", details: nil))
                return
            }
            audioPlayer?.volume = Float(volume)
            result(nil)

        case "setSpeed":
            guard let args = call.arguments as? [String: Any],
                  let speed = args["speed"] as? Double else {
                result(FlutterError(code: "INVALID_ARGS", message: "speed is required", details: nil))
                return
            }
            audioPlayer?.rate = Float(speed)
            result(nil)

        case "getPosition":
            let pos = Int((audioPlayer?.currentTime ?? 0) * 1000)
            result(pos)

        case "getDuration":
            if let player = audioPlayer {
                result(Int(player.duration * 1000))
            } else {
                result(nil)
            }

        case "getState":
            let state: String
            if audioPlayer == nil {
                state = "idle"
            } else if audioPlayer!.isPlaying {
                state = "playing"
            } else {
                state = "paused"
            }
            result(["state": state])

        case "release":
            releasePlayer()
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Player Methods

    private func handleSetFilePath(_ path: String, result: @escaping FlutterResult) {
        releasePlayer()

        let url = URL(fileURLWithPath: path)
        guard FileManager.default.fileExists(atPath: path) else {
            result(FlutterError(code: "FILE_NOT_FOUND", message: "File not found: \(path)", details: nil))
            return
        }

        do {
            audioPlayer = try AVAudioPlayer(contentsOf: url)
            audioPlayer?.delegate = self
            audioPlayer?.enableRate = true
            audioPlayer?.prepareToPlay()

            sendStateEvent("ready")
            if let duration = audioPlayer?.duration {
                sendEvent("duration", data: ["duration": Int(duration * 1000)])
            }
            result(nil)
        } catch {
            result(FlutterError(code: "SET_FILE_FAILED", message: error.localizedDescription, details: nil))
        }
    }

    private func handlePlay(result: @escaping FlutterResult) {
        guard let player = audioPlayer else {
            result(FlutterError(code: "NOT_READY", message: "Player not ready", details: nil))
            return
        }
        player.play()
        startPositionUpdates()
        sendStateEvent("playing")
        result(nil)
    }

    private func handlePause(result: @escaping FlutterResult) {
        guard let player = audioPlayer else {
            result(FlutterError(code: "NOT_READY", message: "Player not ready", details: nil))
            return
        }
        player.pause()
        stopPositionUpdates()
        sendStateEvent("paused")
        result(nil)
    }

    private func handleStop(result: @escaping FlutterResult) {
        guard let player = audioPlayer else {
            result(nil)
            return
        }
        player.stop()
        player.currentTime = 0
        stopPositionUpdates()
        sendStateEvent("stopped")
        result(nil)
    }

    private func handleSeek(positionMs: Int, result: @escaping FlutterResult) {
        guard let player = audioPlayer else {
            result(FlutterError(code: "NOT_READY", message: "Player not ready", details: nil))
            return
        }
        player.currentTime = Double(positionMs) / 1000.0
        result(nil)
    }

    // MARK: - Position Updates

    private func startPositionUpdates() {
        stopPositionUpdates()
        positionTimer = Timer.scheduledTimer(withTimeInterval: 0.2, repeats: true) { [weak self] _ in
            guard let self = self, let player = self.audioPlayer, player.isPlaying else { return }
            self.sendEvent("position", data: ["position": Int(player.currentTime * 1000)])
        }
    }

    private func stopPositionUpdates() {
        positionTimer?.invalidate()
        positionTimer = nil
    }

    // MARK: - Event Sending

    private func sendStateEvent(_ state: String) {
        sendEvent("state", data: ["state": state])
    }

    private func sendEvent(_ type: String, data: [String: Any]) {
        var event: [String: Any] = ["event": type]
        for (key, value) in data {
            event[key] = value
        }
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?(event)
        }
    }

    // MARK: - Release

    private func releasePlayer() {
        stopPositionUpdates()
        audioPlayer?.stop()
        audioPlayer = nil
    }
}

// MARK: - AVAudioPlayerDelegate

extension AudioPlayerPlugin: AVAudioPlayerDelegate {
    public func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
        stopPositionUpdates()
        sendStateEvent("completed")
    }

    public func audioPlayerDecodeErrorDidOccur(_ player: AVAudioPlayer, error: Error?) {
        sendEvent("error", data: ["error": error?.localizedDescription ?? "Decode error"])
    }
}
